### PR TITLE
feature: Registry management UI using GraphQL

### LINKED ISSUE #1894: Container registry page using GraphQL
using https://github.com/lablup/backend.ai/pull/1471
```[tasklist]
### Tasks
- [ ] https://github.com/lablup/backend.ai/pull/1631
- [ ] https://github.com/lablup/backend.ai-webui/pull/1975
```

### PR DESCRIPTION
**Pull request by yomybaby**
resolve #1894

This PR is depends on 
- https://github.com/lablup/backend.ai/pull/1636
- https://github.com/lablup/backend.ai/pull/1631

The new container registry UI is only displayed if the Manage version is compatible with 23.09.2

image
image

## PR Checklist:

- [ ] CRUD registry.
- [ ] Activate/deactivate registry.
- [ ] Re-scan specific registry (This only works when the WebUI and API endpoint have the same domain).

---

### Summary
🚀🛠️📦

This pull request adds a new feature to the web UI for managing container registries using GraphQL. It also updates some UI components and libraries, improves the error handling and code formatting, and adds a feature flag for checking the backend API compatibility.

> _Oh we're the web UI crew and we've got work to do_
> _We're adding features and components for the `container-registry-gql`_
> _We'll update the packages and the schema too_
> _And we'll heave away on the count of three, hooray!_

### Walkthrough
*  Add, modify, and remove types and mutations for container registries in the GraphQL schema ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R25), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R180-R207), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R307-R310), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R648-R660), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L819-R873), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R974-R975))
*  Add, modify, and remove fields for quota scope size, virtual folder count, and virtual folder size in the GraphQL schema ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L210-L211), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L222-R250), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L268-R297), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L544-L545), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L696-L697), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L707-R751), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L766-R812), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L840-R895), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L1079-R1143))
*  Add fields for sudo session feature and node interface in the GraphQL schema ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R802), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R918), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R953), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R1101), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R1129))
*  Rename and add fields for inference session lifecycle and routing errors in the GraphQL schema ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L395-R432), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74L473-R507), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R1027))
*  Update the version of `antd` and its dependencies in `package.json` and `package-lock.json` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L20-R20), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L97-R99), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L5281-R5330), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15681-R15683), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15727-R15729), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15743-R15745), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15774-R15776), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15791-R15798), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15856-R15861), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15887-R15893), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L15934-R15936), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16031-R16033), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16052-R16054), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16100-R16102), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16109-R16109), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16157-R16162), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16206-R16208), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16220-R16225), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8c82bfd96b743d68691d36a5ebe566d90a8d5862b7617fc8a2038d44b929d9e5L16233-R16240), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-5a3f058f37c9436d08fa97bd30209f67c9a1a61d041b0ce27e48613dfb3a1a10L15-R15))
*  Remove unused packages from `package.json` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-5a3f058f37c9436d08fa97bd30209f67c9a1a61d041b0ce27e48613dfb3a1a10L62-L66))
*  Create new components for container registry management in `ContainerRegistryDeleteConfirmModal.tsx`, `ContainerRegistryEditorModal.tsx`, and `PopConfirmWithInput.tsx` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-ff2d94a52416ad5ab41dac80d7329966e22dfe08c681ce05f4ec3edfe6540d2fR1-R7), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-d0457652e1244b1639d000b64c43fe9562acf4dd87f2eed3b06c86a6a61675fdR1-R298), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-e6e3a8684476d6af90ff03182ac2b37c8f0b0c0173a3d2ba82f45dedab874272R1-R102))
*  Create a new custom hook for error message translation in `usePainKiller.tsx` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-fad998151e076953970d2ebe72e266dbfb2056495ac84f1e49594b2b0fab1204R1-R92))
*  Add import statements and custom element definitions for new components in `index.tsx` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8e8a93d52a7e6e8a4b91770c2de7cb6b500148e2855dc2f5bca6f97f8271bc06R3), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8e8a93d52a7e6e8a4b91770c2de7cb6b500148e2855dc2f5bca6f97f8271bc06L52-R57), [link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8e8a93d52a7e6e8a4b91770c2de7cb6b500148e2855dc2f5bca6f97f8271bc06R235-R247))
*  Add conditional rendering logic for `ContainerRegistryList` component in `backend-ai-environment-view.ts` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-2503e01eb2ae708420cffaeb25d58f2d64a81b6e9a02df491a176420493ad56eL184-R195))
*  Add a new feature flag for container registry GraphQL schema in `backend.ai-client-esm.ts` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-788509bd6c57c37a3f54c3b1c52833fb6918816292fc60f5bf08c293e72818a9R663-R665))
*  Add a blank line to the end of `schema.graphql` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-0678c75a543d7ad98fe2396a3584fedcd5cfa3a0e6169cd460ad41d928fd7a74R1188))
*  Remove a blank line in `index.tsx` ([link](https://github.com/lablup/backend.ai-webui/pull/1975/files?diff=unified&w=0#diff-8e8a93d52a7e6e8a4b91770c2de7cb6b500148e2855dc2f5bca6f97f8271bc06L42))

### CODE REVIEW

File: src/components/backend-ai-environment-view.ts [hunk 1]
```diff
@@ -181,9 +181,18 @@ export default class BackendAIEnvironmentView extends BackendAIPage {
...
+              ? html`
+                  <div class="flex" style="height:calc(100vh - 183px);">
+                    <backend-ai-react-container-registry-list></backend-ai-react-container-registry-list>
+                  </div>
+                `
+              : html`
+                  <backend-ai-registry-list
+                    ?active="${this._activeTab === 'registry-lists'}"
+                  ></backend-ai-registry-list>
+                `}
```
Comment (yomybaby) [inline src/components/backend-ai-environment-view.ts:195]: for backwards compatibility, 
- `backend-ai-registry-list`: the previous version which uses `config` REST API
- `backend-ai-react-container-registry-list`: the new version which uses GraphQL

**Comment by yomybaby** [comment]
I'll update `schema.graphql` after the PRs that related to this.

**Review by agatha197 — CHANGES_REQUESTED** [review]
I found an error.
1. create a new registry
2. set password without username.
3. if refresh ->  that registry is deleted.
    if modify -> error occurred.

---
Can you add a warning message? If a user enters a harbor project type that requires an ID and password, such as `testing`, in cr.backend.ai, but there is no ID and password information, that occurs an unexpected error.

---
Let's add a `contextHolder` to message component to avoid a warning message.
image

ref
- antd: https://ant.design/components/message
- control panel
  - https://github.com/lablup/backend.ai-control-panel/pull/683/files#diff-85da2aab8b5207bb4d1a51c90becd03773c4b6c176f2fc87a6e008d046ebf318R29
  - https://github.com/lablup/backend.ai-control-panel/pull/683/files#diff-85da2aab8b5207bb4d1a51c90becd03773c4b6c176f2fc87a6e008d046ebf318R72
  - https://github.com/lablup/backend.ai-control-panel/pull/683/files#diff-85da2aab8b5207bb4d1a51c90becd03773c4b6c176f2fc87a6e008d046ebf318R208

**Comment by yomybaby** [comment]
I update below:
- If `hobor` is selected without id/pw, show confirm when click saving.
- if a password exists, username rule has a required option
  image
- use `useMessage` (because `App` doesn't work properly)

**Review by agatha197 — CHANGES_REQUESTED** [review]
Cannot create a new registry this way.
1. Fill hostname, registry URL, password
2. Click Add button
3. remove input string of password
4. Click Add button
5. `Error Occurred` notification is shown.

**Review by agatha197 — APPROVED** [review]
LGTM!

**Pull request merged**